homeassistant/components/tellduslive/__init__.py

```diff
@@ -4,250 +4,174 @@
-                         'Requesting Telldus Local API configuration')
-            request_configuration(config.get(DOMAIN).get(CONF_HOST))
 
+async def async_setup(hass, config):
+    """Set up the Telldus Live component."""
+    if DOMAIN not in config:
         return True
-    else:
-        _LOGGER.info('Tellstick discovered, awaiting discovery callback')
-        return True
-
-    if not session.is_authorized:
-        _LOGGER.error('Authentication Error')
-        return False
 
-    client = TelldusLiveClient(hass, config, session)
-    hass.data[DOMAIN] = client
-    client.update()
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+            data={
+                KEY_HOST: config[DOMAIN].get(CONF_HOST),
+                KEY_SCAN_INTERVAL: config[DOMAIN].get(CONF_UPDATE_INTERVAL),
+            }))
+    return True
 
-    interval = config.get(DOMAIN, {}).get(CONF_UPDATE_INTERVAL,
-                                          DEFAULT_UPDATE_INTERVAL)
-    _LOGGER.debug('Update interval %s', interval)
-    track_time_interval(hass, client.update, interval)
 
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    interval_tracker = hass.data.pop(INTERVAL_TRACKER)
+    interval_tracker()
+    await asyncio.wait([
+        hass.config_entries.async_forward_entry_unload(config_entry, component)
+        for component in hass.data.pop(CONFIG_ENTRY_IS_SETUP)
+    ])
+    del hass.data[DOMAIN]
+    del hass.data[DATA_CONFIG_ENTRY_LOCK]
     return True
 
 
 class TelldusLiveClient:
     """Get the latest data and update the states."""
 
-    def __init__(self, hass, config, session):
+    def __init__(self, hass, config_entry, session):
         """Initialize the Tellus data object."""
         self._known_devices = set()
 
         self._hass = hass
-        self._config = config
+        self._config_entry = config_entry
         self._client = session
 
-    def update(self, *args):
-        """Update local list of devices."""
+    @staticmethod
+    def identify_device(device):
+        """Find out what type of HA component to create."""
+        if device.is_sensor:
+            return 'sensor'
+        from tellduslive import (DIM, UP, TURNON)
+        if device.methods & DIM:
+            return 'light'
+        if device.methods & UP:
+            return 'cover'
+        if device.methods & TURNON:
+            return 'switch'
+        if device.methods == 0:
+            return 'binary_sensor'
+        _LOGGER.warning("Unidentified device type (methods: %d)",
+                        device.methods)
+        return 'switch'
+
+    async def _discover(self, device_id):
+        """Discover the component."""
+        device = self._client.device(device_id)
+        component = self.identify_device(device)
+        config_entries_key = '{}.{}'.format(component, DOMAIN)
```

Comment:
This is what's in hass.data at `CONFIG_ENTRY_IS_SETUP`. We can't use that directly when unloading the entries.

Reply:
It would make more sense to just store the component in `config_entries_key`..

Reply:
Or what do you think?

Reply:
Yeah, I guess that would be enough. :+1: